packages/devkit/src/generators/project-name-directory-utils.ts

```diff
@@ -0,0 +1,278 @@
+  directory?: string;
+  importPath?: string;
+  nameDirectoryFormat?: ProjectNameDirectoryFormat;
+  rootProject?: boolean;
+};
+
+export type ProjectNamesAndDirectories = {
+  /**
+   * Normalized full project name, including scope if name was provided with
+   * scope (e.g., `@scope/name`, only available when `nameDirectoryFormat` is
+   * `as-provided`).
+   */
+  projectName: string;
+  /**
+   * Normalized project directory, including the layout directory if configured.
+   */
+  projectDirectory: string;
+  names: {
+    /**
+     * Normalized project name without scope. It's meant to be used when
+     * generating file names that contain the project name.
+     */
+    projectFileName: string;
+    /**
+     * Normalized project name without scope or directory. It's meant to be used
+     * when generating shorter file names that contain the project name.
+     */
+    projectSimpleName: string;
+  };
+  /**
+   * Normalized import path for the project.
+   */
+  importPath?: string;
+};
+
+type ProjectNameDirectoryFormats = {
+  'as-provided': {
+    description: string;
+    options: ProjectNamesAndDirectories;
+  };
+  derived?: {
+    description: string;
+    options: ProjectNamesAndDirectories;
+  };
+};
+
+export async function determineProjectNamesAndDirectories(
+  tree: Tree,
+  options: ProjectGenerationOptions
+): Promise<ProjectNamesAndDirectories> {
+  validateName(options.name, options.nameDirectoryFormat);
+  const formats = getProjectNameDirectoryFormats(tree, options);
+  const format =
+    options.nameDirectoryFormat ?? (await determineFormat(formats));
+
+  return formats[format].options;
+}
+
+function validateName(
+  name: string,
+  nameDirectoryFormat?: ProjectNameDirectoryFormat
+): void {
+  if (nameDirectoryFormat === 'derived' && name.startsWith('@')) {
+    throw new Error(
+      `The project name "${name}" cannot start with "@" when the "nameDirectoryFormat" is "derived".`
+    );
+  }
+
+  /**
+   * Matches two types of project names:
+   *
+   * 1. Valid npm package names (e.g., '@scope/name' or 'name').
+   * 2. Names starting with a letter and can contain any character except whitespace and ':'.
+   *
+   * The second case is to support the legacy behavior (^[a-zA-Z].*$) with the difference
+   * that it doesn't allow the ":" character. It was wrong to allow it because it would
+   * conflict with the notation for tasks.
+   */
+  const pattern =
+    '(?:^@[a-zA-Z0-9-*~][a-zA-Z0-9-*._~]*\\/[a-zA-Z0-9-~][a-zA-Z0-9-._~]*|^[a-zA-Z][^:]*)$';
+  const validationRegex = new RegExp(pattern);
+  if (!validationRegex.test(name)) {
+    throw new Error(
+      `The project name should match the pattern "${pattern}". The provided value "${name}" does not match.`
+    );
+  }
+}
+
+async function determineFormat(
+  formats: ProjectNameDirectoryFormats
+): Promise<ProjectNameDirectoryFormat> {
+  if (!formats.derived) {
+    return 'as-provided';
+  }
+
+  if (process.env.NX_INTERACTIVE !== 'true') {
```

Comment:
What would you think about us setting an env variable inside `generate.ts`, along the lines of `NX_ACTIVE_GENERATOR`... Then, when we ask would prompt we could check if that matched our expectation, and if it didn't we would know that it was being ran as part of a community plugin that wrapped us instead of directly, and we could not prompt to be less breaking?

Reply:
I'm achieving the same by having the generator public function (the one exported by the plugin public API and currently consumed by third-party plugins), default to `derived`: https://github.com/nrwl/nx/pull/18420/files#diff-25f2d12e60c477f512e3c94ce788aba9cf83f6f0dd592f3a6ab6cbf7c35452fcR45-R55. Any plugin consuming the library generator won't have its behavior affected. This applies to all, first and third-party plugins. In the case of the first-party plugins, we'll amend them in follow-up PRs.

Last week, I mentioned to @FrozenPandaz the same solution you're proposing (I was thinking of `NX_INITIAL_GENERATOR`, but the idea was the same), but Jason proposed the current approach and we went with that. The environment variable approach would require fewer changes to our generators (we wouldn't need to have two different implementations, one for composition with a default value and another one for the collection). @FrozenPandaz thoughts?

Reply:
We discussed this offline and decided to leave it they way is currently implemented in this PR. It's an idiomatic approach currently supported by our generators.